xgcm/autogenerate.py

```diff
@@ -0,0 +1,240 @@
+        Dict with information on the dimension in ds corrsponding to the xgcm
+        axis. E.g. {'X':'lon'}
+    pos_from : str
+        Position of the gridpoints given in 'ds'. Can be
+        {'center','left','right'}
+    pos_to : str
+        Position of the gridpoints to be generated. Can be
+        {'center','left','right'}
+    wrap : None or float
+        If specified, marks the value of discontinuity across boundary, e.g.
+        360 for global longitude values and 180 for global latitudes.
+    pad : {None, float, 'auto'}
+        If specified, determines the padding to be applied across boundary.
+        If float is specified, that value is used as padding. Auto attempts to
+        pad linearly extrapolated values. Can be useful for e.g. depth
+        coordinates (to reconstruct 0 depth). Can lead to unexpected values
+        when coordinate is multidimensional.
+    raw_switch : bool
+        Determines if the attributes are created from scratch. Should be
+        enabled for dimensions and deactivated for multidimensional
+        coordinates. These can only be calculated after the dims are created.
+    """
+    new_name = name+'_inferred'
+
+    # Determine the relative position to interpolate to based on current and
+    # desired position
+    relative_pos_to = position_to_relative(pos_from, pos_to)
+
+    if (wrap is not None) and (pad is not None):
+        raise RuntimeError('Coordinate cannot be wrapped and padded at the\
+                            same time')
+    elif (wrap is None) and (pad is not None):
+        if pad == 'auto':
+            left_pad, right_pad = auto_pad(ds[name], axis_dim)
+            if relative_pos_to == 'right':
+                fill_value = right_pad
+            elif relative_pos_to == 'left':
+                fill_value = left_pad
+        else:
+            fill_value = pad
+        periodic = False
+        boundary = 'fill'
+    elif (wrap is not None) and (pad is None):
+        periodic = True
+        fill_value = 0.0
+        boundary = None
+    else:
+        raise RuntimeError('Either "wrap" or "pad" have to be specified')
+
+    if raw_switch:
+        # Input coordinate has to be declared as center,
+        # or xgcm.Axis throws error. Will be rewrapped below.
+        ds[name] = fill_attrs(ds[name], 'center', axis)
+
+        ax = Axis(ds, axis, periodic=periodic, wrap=wrap, raw_out=True)
+        ds.coords[new_name] = ax.interp(ds[name], relative_pos_to,
+                                        boundary=boundary,
+                                        fill_value=fill_value)
+
+        # Place the correct attributes
+        ds[name] = fill_attrs(ds[name], pos_from, axis)
+        ds[new_name] = fill_attrs(ds[new_name], pos_to, axis)
+    else:
+        ax = Axis(ds, axis, periodic=periodic, wrap=wrap, raw_out=False)
+        ds.coords[new_name] = ax.interp(ds[name], pos_to, boundary=boundary,
+                                        fill_value=fill_value)
```

Comment:
Can you help me understand what is happening here? What is `raw_switch`? What do these two cases represent?

Reply:
Ok so for a set of coordinates there are two fundamental cases. The coordinates are a) one dimensional (dimensions) or 2) multidimensional. These are separated by the keyword  raw_switch.

These two cases are treated differently because for each dataset we need to recreate all a) cases before we can proceed to 2), hence this is really the 'raw' data processing step. If we have working one dimensional coordinates (e.g. after we looped over the `axes_dims_dict`, we can use the regular xgcm.Axis to interpolate multidimensional coordinates. 
I designed it this way so that any changes to the Axis.interp method can directly propagate to this module.

Does that make any sense?

Reply:
Now I understand and it makes sense.

That would have been a good comment to include in the code.

Reply:
It is included now